README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis build
status](https://travis-ci.org/kylebittinger/abdiv.svg?branch=master)](https://travis-ci.org/kylebittinger/abdiv)

# abdiv

This package re-implements measures of ecological diversity from several
other software packages, including `vegan`, `scikit-bio`, `Mothur`, and
`GUniFrac`.

## Installation

You can install `abdiv` from github with:

``` r
# install.packages("devtools")
devtools::install_github("kylebittinger/abdiv")
```

## Alpha diversity

Let’s say we’ve surveyed a field and counted the number of plants in
each of two sites. We’ve found five species in total, and we’d like to
summarize the diversity of the two sampling sites. The diversity within
each site is called α-diversity.

Here are the number of plants for each species at site 1 and site 2,
represented as vectors.

``` r
site1 <- c(2, 5, 16, 0, 1)
site2 <- c(0, 0, 8, 8, 8)
```

The two sites have about the same number of total plants, but the
distribution of species is much different. More than half the plants in
site 1 belong to a single species, whereas the plants in site 2 are
almost evenly distributed across three different species. To get
started, let’s look at a few ways to quantify the α-diversity for each
sample.

Richness measures the total number of species in each sample.

``` r
richness(site1)
## [1] 4
richness(site2)
## [1] 3
```

The Shannon index measures both the number of species and the evenness
of the relative abundance values. Site 2 has fewer species, but each
species has the same relative abundance, so the Shannon index is higher.

``` r
shannon(site1)
## [1] 0.9365995
shannon(site2)
## [1] 1.098612
```

Let’s summarize the diversity of site 1 and site 2 using all the
functions available in this library. The full set of within-sample
diversity functions is available as a character vector in
`alpha_diversities`. The term “α-diversity” means the diversity within a
single sample.

``` r
library(tidyverse)
tibble(Site = rep(c("Site 1", "Site 2"), each=5), Counts = c(site1, site2)) %>%
  group_by(Site) %>%
  summarize_at(vars(Counts), alpha_diversities) %>%
  pivot_longer(-Site, names_to = "Measure") %>%
  ggplot(aes(x=Measure, y=value, color=Site)) +
  geom_point() +
  scale_color_manual(values=c("#E64B35", "#4DBBD5")) +
  coord_flip() +
  theme_bw()
```

![](tools/readme/alpha-diversity-1.png)<!-- -->

We can see that site 1 is regarded as more diverse by some measures; it
has the most species. For other measures, site 2 is regarded as more
diverse; it has the most even distribution of species.

In our documentation, you can find more info on each α-diversity
function.

## Beta diversity

Having assessed the diversity within each sample, we can next ask about
the number of shared species between sites. If species are shared, how
similar is the distribution across species? There are many ways to
quantify the between-sample diversity or β-diversity.

You can think about β-diversity as either the similarity or
dissimilarity between sites. The functions in `abdiv` are written in
terms of dissimilarity: similar sites will have values close to zero,
and highly dissimilar sites will have values close to the maximum. This
way of thinking goes along with our intuition about diversity: sites
with greater dissimilarity will exhibit increased diversity if we
consider both sites together.

Let’s look at some examples. The Jaccard distance counts the fraction of
species present in only one site. The answer is 3 out of 5, or 0.6.

``` r
jaccard(site1, site2)
## [1] 0.6
```

The Bray-Curtis dissimilarity adds up the absolute differences between
species counts, then divides by the total counts. For our two sites,
that’s (2 + 5 + 8 + 8 + 7) / 48, or 0.625.

``` r
bray_curtis(site1, site2)
## [1] 0.625
```

Again, we’ll use a vector called `beta_diversities` to compute every
dissimilarity measure in the library.

``` r
tibble(Measure = beta_diversities) %>%
  group_by(Measure) %>%
  mutate(value = get(Measure)(site1, site2)) %>%
  ggplot(aes(x=Measure, y=value)) +
  geom_point(color="#4DBBD5") +
  scale_y_log10() +
  coord_flip() +
  theme_bw()
```

![](tools/readme/beta-diversity-1.png)<!-- -->

The dissimilarities are generally positive, and they have a range of
scales. Some dissimilarity measures range from 0 to 1, while others can
go up indefinitely.

As before, you can find more info on each β-diversity function in our
documentation.

## Phylogenetic diversity and UniFrac

A phylogenetic tree can be incorporated into diversity measures to add
information about the evolutionary history of species. One of the first
measures in this area was Faith’s phylogenetic diversity. Here, we take
an example from Figure 1 in Faith and Richards (PMID 24832524) to show
how phylogenetic diversity works. The tree is included as `faith_tree`:

``` r
library(ggtree)
ggtree(faith_tree, ladderize = F) +
  geom_tiplab()
```

<img src="tools/readme/unnamed-chunk-7-1.png" style="display: block; margin: auto;" />

If all the species are present, the value of Faith’s phylogenetic
diversity (PD) is the sum of the branch lengths. Here, we expect the
total branch length to be 5 + 4 + 2 + 4 + 1 + 20 + 5 + 1 + 3 = 45,
adding from top to bottom.\[1\]

Here is how you would calculate Faith’s PD:

``` r
faith_pd(c(1, 1, 1, 1, 1), faith_tree)
## [1] 45
```

If species “a” is missing, we expect Faith’s PD to be reduced by 5, the
length of the branch leading to species “a”.

``` r
faith_pd(c(0, 1, 1, 1, 1), faith_tree)
## [1] 40
```

The practice of using phylogenetic information in diversity has been
especially popular in microbial ecology, where bacteria are surveyed
using the 16S rRNA marker gene. In addition to serving as a fingerprint
for bacteria, the gene sequence can be used to build a phylogenetic
tree.

In this area, the UniFrac distance is widely used to measure β-diversity
between bacterial communities. We’ll reproduce an example from the
UniFrac paper by Lozupone and Knight (PMID 16332807), which describes
the unweighted UniFrac distance. The tree from Figure 1 is available as
`lozupone_tree`.

In the example from Figure 1A, we’ve measured two bacterial communities,
where the species detected in each are labeled with circles and squares.
The communities have no species in common.

``` r
ggtree(lozupone_tree, ladderize = F) %<+%
  lozupone_panel_a +
  geom_tippoint(aes(shape=SampleID), x=2.6, size=3) +
  scale_shape_manual(values = c(1, 15)) +
  scale_x_continuous(limits=c(0, 2.8))
```

<img src="tools/readme/unnamed-chunk-10-1.png" style="display: block; margin: auto;" />

The circle and square communities are all mixed up on the phylogenetic
tree. The unweighted UniFrac distance is the fraction of the total
branch length where the branch leads to a circle or square, but not
both. Here, we would only count the branches leading to the tips of the
tree. The length of these branches, added together, is about half the
total length for all branches in the tree.

``` r
a_circle <- c(1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1)
a_square <- c(0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0)
unweighted_unifrac(a_circle, a_square, lozupone_tree)
## [1] 0.5378151
```

We can increase the UniFrac distance if we rearrange the communities so
that all the circles are in the upper part of the tree and all the
squares are in the lower part. In the paper, this is shown in Figure 1B.

``` r
ggtree(lozupone_tree, ladderize = F) %<+%
  lozupone_panel_b +
  geom_tippoint(aes(shape=SampleID), x=2.6, size=3) +
  scale_shape_manual(values = c(1, 15)) +
  scale_x_continuous(limits=c(0, 2.8))
```

<img src="tools/readme/unnamed-chunk-12-1.png" style="display: block; margin: auto;" />

With this arrangement, the species in each community are
phylogenetically very different. Except for the root, each branch of the
tree leads uniquely to a species present in square or circle, and never
to a species present in both. Therefore, the UniFrac distance is close
to 1.

``` r
b_circle <- c(1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0)
b_square <- c(0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1)
unweighted_unifrac(b_circle, b_square, lozupone_tree)
## [1] 0.9747899
```

## Practical use

Now that we’ve introduced you to the functions in this package, we
should probably tell you how to use them in your work. Your data are
likely to be in one of two formats: a long-format data frame, or a
matrix of counts. We’ll tackle the long format first.

First, we’ll arrange the data from the plant survey example in long
format. We’ll also add a third site to help with the demonstration.

``` r
site3 <- c(15, 1, 4, 2, 2)
plants <- tibble(
  Site = rep(c("Site 1", "Site 2", "Site 3"), each = 5),
  Species = rep(letters[1:5], times = 3),
  Counts = c(site1, site2, site3)
)
plants
## # A tibble: 15 x 3
##    Site   Species Counts
##    <chr>  <chr>    <dbl>
##  1 Site 1 a            2
##  2 Site 1 b            5
##  3 Site 1 c           16
##  4 Site 1 d            0
##  5 Site 1 e            1
##  6 Site 2 a            0
##  7 Site 2 b            0
##  8 Site 2 c            8
##  9 Site 2 d            8
## 10 Site 2 e            8
## 11 Site 3 a           15
## 12 Site 3 b            1
## 13 Site 3 c            4
## 14 Site 3 d            2
## 15 Site 3 e            2
```

If we want to compute α-diversity, we can group by `Site` and create a
new column with the answer. The α-diversity functions take the number of
counts for each species as an argument.

``` r
plants %>%
  group_by(Site) %>%
  summarize(Richness = richness(Counts))
## # A tibble: 3 x 2
##   Site   Richness
##   <chr>     <int>
## 1 Site 1        4
## 2 Site 2        3
## 3 Site 3        5
```

If you want to cover more than one α-diversity measure, you can use
`summarize_at()`.

``` r
plants %>%
  group_by(Site) %>%
  summarise_at(vars(Counts), c("shannon", "invsimpson"))
## # A tibble: 3 x 3
##   Site   shannon invsimpson
##   <chr>    <dbl>      <dbl>
## 1 Site 1   0.937       2.01
## 2 Site 2   1.10        3   
## 3 Site 3   1.14        2.30
```

The old school way to compute α-diversity is to arrange your data in a
matrix format, with rows representing different sites or “observations”,
and columns representing species. This is how you would do it with the
`vegan` package, for example. Here is our data in matrix format:

``` r
plants_matrix <- matrix(
  c(site1, site2, site3), nrow=3, byrow=TRUE,
  dimnames = list(c("Site 1", "Site 2", "Site 3"), letters[1:5]))
plants_matrix
##         a b  c d e
## Site 1  2 5 16 0 1
## Site 2  0 0  8 8 8
## Site 3 15 1  4 2 2
```

In matrix format, we can use the `apply()` function from base R to get
the diversity for each site.

``` r
apply(plants_matrix, 1, shannon)
##    Site 1    Site 2    Site 3 
## 0.9365995 1.0986123 1.1389489
```

Here is the same example using `vegan`.

``` r
vegan::diversity(plants_matrix)
##    Site 1    Site 2    Site 3 
## 0.9365995 1.0986123 1.1389489
```

The long format is friendlier to workflows using functions in the
`tidyverse`, but the matrix format has some advantages. The matrix
format ensures that every sample has a valid count value for every
species considered; the matrix contains values of zeros for species not
observed. This quality is not critically important for alpha-diversity
measures, but it is important for computing β-diversity.

For β-diversity, we recommend proceeding via the matrix format. If your
data is in long format, the `usedist` package has a function to convert
to a numeric matrix.\[2\]

``` r
usedist::pivot_to_numeric_matrix(plants, Site, Species, Counts)
##         a b  c d e
## Site 1  2 5 16 0 1
## Site 2  0 0  8 8 8
## Site 3 15 1  4 2 2
```

To create a distance matrix between samples, we can employ the
`dist_make()` function from `usedist`.

``` r
usedist::dist_make(plants_matrix, bray_curtis)
##           Site 1    Site 2
## Site 2 0.6250000          
## Site 3 0.6666667 0.6666667
```

## Practical use - phylogenetic diversity

One practical issue with phylogenetic diversity is to make sure that the
order of the species in your vectors matches the order of the tree. If
your vector is not named, and you’re sure that the order matches, you
can use the phylogenetic diversity functions as above. To take the
example from Faith’s phylogenetic diversity, let’s say we had a vector
representing 10 counts of species “d” and 12 of species “e”.

``` r
faith_pd(c(0, 0, 0, 10, 12), faith_tree)
## [1] 29
```

If your vector is named, the names will be automatically used to match
the vector with the tree. Missing names are filled in with zero counts.

``` r
faith_pd(c(d=10, e=12), faith_tree)
## [1] 29
```

If your vector is not named, or if you’re not totally sure that the
order of species in your vector matches the order of species in the
tree, you can pass in the species labels for your vector as an
additional argument.

``` r
faith_pd(c(10, 12), faith_tree, c("d", "e"))
## [1] 29
```

This last approach is useful for a `tidyverse` workflow, where the
vectors are not named and the species labels are found in a separate
column. Here is an example for data in long format, where the species
labels are in the column, “Species”:

``` r
plants %>%
  group_by(Site) %>%
  summarize(FaithPD = faith_pd(Counts, faith_tree, Species))
## # A tibble: 3 x 2
##   Site   FaithPD
##   <chr>    <dbl>
## 1 Site 1      40
## 2 Site 2      38
## 3 Site 3      45
```

The rules are the same for phylogenetic β-diversity functions. The
`dist_make()` function from `usedist` will pass along additional keyword
arguments to the distance function, so you can give the tree and species
labels to `dist_make()`. Here’s an example from the plants matrix.

``` r
usedist::dist_make(plants_matrix, unweighted_unifrac, faith_tree)
##           Site 1    Site 2
## Site 2 0.2666667          
## Site 3 0.1111111 0.1555556
```

Now, we’ll re-order the species in the data matrix to see how things
work.

``` r
species_reorder <- c("d", "e", "a", "b", "c")
plants_matrix_reorder <- plants_matrix[,species_reorder]
plants_matrix_reorder
##        d e  a b  c
## Site 1 0 1  2 5 16
## Site 2 8 8  0 0  8
## Site 3 2 2 15 1  4
```

The column names of the matrix are automatically added to vectors
extracted from the matrix, so we get the same result as
before.

``` r
usedist::dist_make(plants_matrix_reorder, unweighted_unifrac, faith_tree)
##           Site 1    Site 2
## Site 2 0.2666667          
## Site 3 0.1111111 0.1555556
```

If the column names are missing, first of all, you’re probably in
trouble. But if you do have the species labels stored in a separate
vector, you can pass them along to the distance function. Let’s remove
the column names from our matrix to check this out.

``` r
plants_matrix_reorder_nonames <- plants_matrix_reorder
colnames(plants_matrix_reorder_nonames) <- NULL
plants_matrix_reorder_nonames
##        [,1] [,2] [,3] [,4] [,5]
## Site 1    0    1    2    5   16
## Site 2    8    8    0    0    8
## Site 3    2    2   15    1    4
```

Now, we can pass the species labels to `dist_make()` and verify that we
get the correct answer.

``` r
usedist::dist_make(
  plants_matrix_reorder_nonames, unweighted_unifrac, faith_tree,
  species_reorder)
##           Site 1    Site 2
## Site 2 0.2666667          
## Site 3 0.1111111 0.1555556
```

## Support

Please don’t hesitate to reach out via email or file an issue if you
need support when using this library.

## Footnotes

1.  The answer here is slightly different than that in the paper. See
    `faith_tree` documentation for further explanation.

2.  Full disclosure: `usedist` and `abdiv` are authored by the same
    person.


NEWS.md
# abdiv 0.3.0


# abdiv 0.2.0

* First CRAN release.
* Added many new functions.
* Documented and tested everything.
* Wrote a long README.

Notes on correspondence between this implementation and other software can be
found in comments at the bottom of R/alpha.R and R/beta.R.


R/unifrac.R
# Matrix of branches (rows) vs. leaves (columns)
# Can be used to find the nodes to add up for each branch.
make_edge_matrix <- function (tree) {
  UseMethod("make_edge_matrix")
}

make_edge_matrix.phylo <- function (tree) {
  edge_nodes <- tree$edge[,2]
  np <- ape::nodepath(tree)
  do.call(cbind, lapply(np, function (x) as.integer(edge_nodes %in% x)))
}

make_edge_matrix.nptree <- function (tree) {
  tree$nodepath_matrix
}

cache_nodepath <- function (phylo_tree) {
  phylo_tree$nodepath_matrix <- make_edge_matrix(phylo_tree)
  class(phylo_tree) <- c("nptree", class(phylo_tree))
  phylo_tree
}

get_branch_abundances <- function (edge_matrix, node_abundances) {
  as.numeric(edge_matrix %*% node_abundances)
}

check_tree <- function (tree) {
  if (is.null(tree$edge.length)) {
    stop("Tree has no branch lengths")
  }
}

#' Match vector of counts to phylogenetic tree
#'
#' @param x A vector of species counts.
#' @param tree A phylogenetic tree of class \code{"phylo"}.
#' @param x_labels A vector of species labels for \code{x}.
#' @return The vector \code{x}, re-arranged to match the tree.
#' @details This function applies a couple of different methods to arrange the
#'   data in \code{x} to match a phylogenetic tree.
#'   \itemize{
#'     \item If \code{x_labels} is provided, we use this vector to match the
#'       elements of \code{x} with the tip labels in the tree.
#'     \item If \code{x_labels} is not provided and \code{x} is a named vector,
#'       we use the names to match the tip labels in the tree.
#'     \item If \code{x_labels} is not provided and \code{x} is not named, we
#'       assume that \code{x} is already in the correct order, check that the
#'       length of \code{x} matches the number of tips in the tree, and return
#'       \code{x}.
#'   }
#' @export
match_to_tree <- function (x, tree, x_labels = NULL) {
  tree_labels <- tree$tip.label
  x_labels_are_provided <- !is.null(x_labels)
  if (x_labels_are_provided) {
    if (length(x) != length(x_labels)) {
      stop("Length of x does not match length of x_labels.")
    }
    if (!all(x_labels %in% tree_labels)) {
      x_labels_not_in_tree <- setdiff(x_labels, tree_labels)
      stop(paste("x_labels not found in tree:", x_labels_not_in_tree))
    }
    idx <- match(x_labels, tree_labels)
    xout <- numeric(length(tree_labels))
    xout[idx] <- x
    return(xout)
  }
  x_has_names <- !is.null(names(x))
  if (x_has_names) {
    if (!all(names(x) %in% tree_labels)) {
      x_names_not_in_tree <- setdiff(names(x), tree_labels)
      stop(paste("Names of x not found in tree:", x_names_not_in_tree))
    }
    idx <- match(names(x), tree_labels)
    xout <- numeric(length(tree_labels))
    xout[idx] <- x
    return(xout)
  }
  if (length(x) != length(tree_labels)) {
    stop("Length of x does not match number of tips in tree.")
  }
  return(x)
}

#' Faith's phylogenetic diversity
#'
#' Faith's phylogenetic diversity gives the total branch length on a
#' phylogenetic tree that is spanned by a community.  The abundance of each
#' species in the community is not considered.
#'
#' @param x A numeric vector of species counts or proportions, or a logical
#'   vector of species presence/absence.
#' @param tree A phylogenetic tree object..
#' @param x_labels A character vector of species labels for \code{x}.
#' @return Faith's phylogenetic diversity, \eqn{PD \geq 0}.
#' @details If the vector \code{x} is named, the names will be automatically
#'   used to match \code{x} with the tree. Missing names are filled in with
#'   zero counts. If \code{x} is not named and \code{x_labels} is provided,
#'   these labels are used to match the elements of \code{x} with the tree.
#'   If \code{x} is not named and \code{x_labels} is not provided, it is
#'   assumed that \code{x} is already in the correct order, and we simply
#'   check that its length matches the number of tips in the tree.
#' @references Faith DP. Conservation evaluation and phylogenetic diversity.
#'   Biol. Conserv. 1992;61:1–10. doi: 10.1016/0006-3207(92)91201-3.
#' @examples
#' # Faith's phylogenetic diversity for whole tree is equal to the sum of the
#' # branch lengths.
#' sum(faith_tree$edge.length)
#' faith_pd(c(1, 1, 1, 1, 1), faith_tree)
#'
#' # Can use named vector or additional argument to match species to tree.
#' faith_tree$tip.label
#' faith_pd(c(0, 0, 0, 10, 12), faith_tree)
#' faith_pd(c(d=10, e=12), faith_tree)
#' faith_pd(c(10, 12), faith_tree, c("d", "e"))
#' @export
faith_pd <- function (x, tree, x_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, x_labels)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x) > 0
  sum(b[px])
}

#' UniFrac distance
#'
#' The UniFrac distance is a phylogenetically-weighted distance between two
#' communities of organisms. The measure has been extended a number of
#' times to include abundance-weighted and variance-adjusted versions.
#'
#' @param x,y Numeric vectors of species counts or proportions.
#' @param tree A phylogenetic tree object.
#' @param xy_labels A character vector of species labels for \code{x} and
#'   \code{y}.
#' @param alpha Generalized UniFrac parameter.
#' @return The UniFrac distance between communities \code{x} and \code{y}. The
#'   distance is not defined if either \code{x} or \code{y} have all zero
#'   elements. We return \code{NaN} if this is the case.
#' @details
#' These functions compute different variations of the UniFrac distance between
#' communities described by the vectors \code{x} and \code{y}. If the vectors
#' are named, the names will be automatically used to match the vectors with
#' the tree. Missing names are filled in with zero counts. If the vectors are
#' not named and \code{xy_labels} is provided, these labels will be used to
#' match the vectors with the tree. If the vectors are not named and
#' \code{xy_labels} is not provided, it is assumed that the vectors are already
#' in the correct order, and we simply check that their length matches the
#' number of tips in the tree.
#'
#' \code{unweighted_unifrac} gives the original UniFrac distance from Lozupone
#' and Knight (2005), which is the fraction of total branch length leading to
#' community \code{x} or community \code{y}, but not both. It is based on
#' species presence/absence.
#'
#' \code{weighted_unifrac} gives the abundance-weighted version of UniFrac
#' proposed by Lozupone et al. (2007). In this measure, the branch lengths of
#' the tree are multiplied by the absolute difference in species abundances
#' below each branch.
#'
#' \code{weighted_normalized_unifrac} provides a normalized version of
#' \code{weighted_unifrac}, so the distance is between 0 and 1.
#'
#' \code{variance_adjusted_unifrac} was proposed by Chang et al. (2011) to
#' adjust for the variation of weights in weighted UniFrac under random
#' sampling.
#'
#' \code{generalized_unifrac} was proposed by Chen et al. (2012) to provide a
#' unified mathematical framework for weighted and unweighted UniFrac distance.
#' It includes a parameter, \eqn{\alpha}, which can be used to adjust the
#' abundance-weighting in the distance. A value of \eqn{\alpha = 1} corresponds
#' to weighted UniFrac. A value of \eqn{\alpha = 0} corresponds to unweighted
#' UniFrac if presence/absence vectors are provided.  The authors suggest a
#' value of \eqn{\alpha = 0.5} as a compromise between weighted and unweighted
#' distances.
#'
#' \code{information_unifrac} was proposed by Wong et al. (2016) to connect
#' UniFrac distance with compositional data analysis. They also proposed a
#' "ratio UniFrac" distance, which is not yet implemented.
#'
#' \code{phylosor}, proposed by Bryant et al. (2008), is closely related to
#' unweighted UniFrac distance. If unweighted UniFrac distance is the analogue
#' of Jaccard distance using branches on a phylogenetic tree, PhyloSor is the
#' analogue of Sorenson dissimilarity.
#' @references
#' Lozupone C, Knight R. UniFrac: a new phylogenetic method for
#' comparing microbial communities. Applied and environmental microbiology.
#' 2005;71(12):8228–8235. 10.1128/AEM.71.12.8228-8235.2005
#'
#' Lozupone CA, Hamady M, Kelley ST, Knight R. Quantitative and
#' qualitative \eqn{\beta} diversity measures lead to different insights into
#' factors that structure microbial communities. Applied and environmental
#' microbiology. 2007;73(5):1576–1585. 10.1128/AEM.01996-06
#'
#' Chang Q., et al. Variance adjusted weighted UniFrac: a powerful
#' beta diversity measure for comparing communities based on phylogeny.
#' BMC Bioinformatics. 2011;12:118. 10.1186/1471-2105-12-118
#'
#' Chen J, Bittinger K, Charlson ES, Hoffmann C, Lewis J, Wu GD,
#' et al. Associating microbiome composition with environmental covariates
#' using generalized UniFrac distances. Bioinformatics.
#' 2012;28(16):2106–2113. 10.1093/bioinformatics/bts342
#'
#' Wong RG, Wu JR, Gloor GB. Expanding the UniFrac Toolbox.
#' PLOS ONE. 2016;11(9):1–20. 10.1371/journal.pone.0161196
#'
#' Bryant JA, Lamanna C, Morlon H, Kerkhoff AJ, Enquist BJ, Green JL.
#' Microbes on mountainsides: contrasting elevational patterns of bacterial
#' and plant diversity. Proc Natl Acad Sci U S A. 2008;105 Suppl 1:11505-11.
#' 10.1073/pnas.0801920105
#' @examples
#' # From Lozupone and Knight (2005), Figure 1.
#' # Panel A
#' x1 <- c(1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1)
#' x2 <- c(0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0)
#' unweighted_unifrac(x1, x2, lozupone_tree)
#'
#' # Panel B
#' x3 <- c(0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1)
#' x4 <- c(1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0)
#' unweighted_unifrac(x3, x4, lozupone_tree)
#'
#' # Can use named vectors to specify species
#' weighted_normalized_unifrac(
#'   c(A=1, C=1, D=1, F=1, I=1, L=1, N=1),
#'   c(B=1, E=1, G=1, H=1, J=1, K=1, M=1),
#'   lozupone_tree)
#' weighted_normalized_unifrac(x1, x2, lozupone_tree)
#'
#' # Generalized UniFrac is equal to weighted normalized UniFrac when alpha = 1
#' generalized_unifrac(x1, x2, lozupone_tree, alpha=1)
#' generalized_unifrac(x1, x2, lozupone_tree, alpha=0.5)
#' @name unifrac
NULL

#' @rdname unifrac
#' @export
unweighted_unifrac <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x) > 0
  py <- get_branch_abundances(em, y) > 0
  # Unique branch length is the sum of branch lengths where
  # either px or py is nonzero, but not both
  branch_unique <- sum(b * xor(px, py))
  # Total branch length is the sum of branch lengths where
  # px, py, or both are nonzero
  branch_total <- sum(b[px | py])
  branch_unique / branch_total
}

#' @rdname unifrac
#' @export
weighted_unifrac <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  x <- x / sum(x)
  y <- y / sum(y)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x)
  py <- get_branch_abundances(em, y)
  sum(b * abs(px - py))
}

#' @rdname unifrac
#' @export
weighted_normalized_unifrac <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  x <- x / sum(x)
  y <- y / sum(y)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x)
  py <- get_branch_abundances(em, y)
  sum(b * abs(px - py)) / sum(b * (px + py))
}

#' @rdname unifrac
#' @export
variance_adjusted_unifrac <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  xy_sum <- x + y
  x <- x / sum(x)
  y <- y / sum(y)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x)
  py <- get_branch_abundances(em, y)
  m <- get_branch_abundances(em, xy_sum)
  # (m_total - m) appears in the denominator; must remove double zeroes now
  m_total <- sum(xy_sum)
  keep <- (m != m_total) & (m > 0)
  b <- b[keep]
  px <- px[keep]
  py <- py[keep]
  m <- m[keep]
  # Typos in paper: this quantity should be square rooted, and the first m
  # should have an i subscript
  m_denom <- sqrt(m * (m_total - m))
  sum(b * abs(px - py) / m_denom) / sum(b * (px + py) / m_denom)
}

#' @rdname unifrac
#' @export
generalized_unifrac <- function (x, y, tree, alpha = 0.5, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  x <- x / sum(x)
  y <- y / sum(y)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x)
  py <- get_branch_abundances(em, y)
  # px + py appears in the denominator; must remove double zeroes now
  p_sum <- px + py
  p_sum_defined_and_zero <- (p_sum == 0) %in% TRUE
  numerator <- ifelse(
    p_sum_defined_and_zero, 0, b * (p_sum ^ (alpha - 1)) * abs(px - py))
  denominator <- ifelse(p_sum_defined_and_zero, 0, b * (p_sum ^ alpha))
  sum(numerator) / sum(denominator)
}

#' @rdname unifrac
#' @export
information_unifrac <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  x <- x / sum(x)
  y <- y / sum(y)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x)
  py <- get_branch_abundances(em, y)
  # log(0) produces infinities; p * log(p) must be zero for formula to work
  plogp <- function (p) ifelse((p == 0) %in% TRUE, 0, p * log(p))
  sum(b * abs(plogp(px) - plogp(py))) / sum(b)
}

#' @rdname unifrac
#' @export
phylosor <- function (x, y, tree, xy_labels = NULL) {
  check_tree(tree)
  x <- match_to_tree(x, tree, xy_labels)
  y <- match_to_tree(y, tree, xy_labels)
  em <- make_edge_matrix(tree)
  b <- tree$edge.length
  px <- get_branch_abundances(em, x) > 0
  py <- get_branch_abundances(em, y) > 0
  # Making this look like the formula in Bryant (2008).
  BL_ij <- sum(b[px & py])
  BL_i <- sum(b[px])
  BL_j <- sum(b[py])
  1 - (BL_ij / (0.5 * (BL_i + BL_j)))
}

#' Example data for Faith's phylogenetic diversity
#'
#' This example was used to illustrate phylogenetic diversity in Faith and
#' Richards (2012).
#'
#' @format \code{faith_tree} is a phylogenetic tree with five tips, labeled
#' a-e. It was created with the \code{ape} library.
#'
#' @details In the paper, they give the total branch length as 41, but they
#' don't assign a length to the branch leading to species "b" and "c". Looking
#' at the figure, we estimated that the length should be 4. For that reason,
#' the total branch length of \code{faith_tree} is 45, rather than 41.
#'
#' @source
#' Faith DP, Richards ZT. Biology (Basel). 2012;1(3):906-32.
#' 10.3390/biology1030906
"faith_tree"

#' Example data for UniFrac distance
#'
#' This example was used to illustrate unweighted UniFrac distance in Lozupone
#' and Knight (2005).
#'
#' @format \code{lozupone_tree} is a phylogenetic tree with 14 tips, labeled
#' A-N. It was created with the \code{ape} library.
#'
#' The data frames \code{lozupone_panel_a} and \code{lozupone_panel_b} are
#' transcribed from Figure 1 of the paper. They have the following columns:
#' \describe{
#'   \item{Species}{The species label, matching to the phylogenetic tree.}
#'   \item{SampleID}{The community, either "Circle" or "Square".}
#'   \item{Counts}{
#'     The number of organisms counted per species, always 1 for this
#'     example.}
#' }
#' @source
#' Lozupone C, Knight R. UniFrac: a new phylogenetic method for
#' comparing microbial communities. Applied and environmental microbiology.
#' 2005;71(12):8228–8235. 10.1128/AEM.71.12.8228-8235.2005
"lozupone_tree"

#' @rdname lozupone_tree
"lozupone_panel_a"

#' @rdname lozupone_tree
"lozupone_panel_b"
